### PR TITLE
fix #33188 MSAA depth buffer not used for external texture

### LINKED ISSUE #54: GLES2 MSAA settings not working
The MSAA settings (Under Project Settings->Rendering->Quality->Msaa seem to have no effect on the actual rendering inside the Oculus Quest in GLES2.

Here is a comparison shot with MSAA off (left) vs MSAA 4x (right):
mass_comparison_off_4x

I tried to investigate briefly; My first suspicion is that in `void RasterizerStorageGLES2::render_target_set_external_texture(RID p_render_target, unsigned int p_texture_id)`: https://github.com/godotengine/godot/blob/1d9233c3882afe888b9396f7f2aac917d4dcac4d/drivers/gles2/rasterizer_storage_gles2.cpp#L5063
no check if msaa is enabled (and no call to `glRenderbufferStorageMultisample` or `glFramebufferTexture2DMultisample` is performed) when the external eye target texture is attached to the framebuffer for rendering.

### LINKED ISSUE #33188: Oculus Mobile VR (Android) MSAA with GLES2 not working
**Godot version:**
Current master (d482a1d8e8b4e68f2f386ba618cd091c79c20424)

**OS/device including version:**
Oculus Quest (firmware version 10, android)

**Issue description:**
Turning on MSAA in GLES2 has no effect in the final rendered image on the Oculus Quest.

The issue was raised in https://github.com/GodotVR/godot_oculus_mobile/issues/54 but initial investigation showed that it might be in the GLES2 renderer and how it handles the external texture from the ARVR interface (around this location maybe: https://github.com/godotengine/godot/blob/d482a1d8e8b4e68f2f386ba618cd091c79c20424/drivers/gles2/rasterizer_storage_gles2.cpp#L5160 as there are no checks for MSAA or framebuffer setup calls that would consider MSAA)

This issue might also be present on DesktopVR headsets when using the GLES2 renderer but I have no hardware to confirm this.

**Steps to reproduce:**
The demo project included https://github.com/GodotVR/godot_oculus_mobile shows the issue when switching MSAA off/on. See also https://github.com/GodotVR/godot_oculus_mobile/issues/54

### PR DESCRIPTION
**Pull request by NeoSpark314**
When rendering to an external texture and MSAA was active (as happened
in the Oculus Mobile ARVR plugin) no MSAA was rendered as the correct
depth buffer and multisample texture target was not used.
This also fixes https://github.com/GodotVR/godot_oculus_mobile/issues/54

*Bugsquad edit:* Fixes #33188.

### CODE REVIEW

File: drivers/gles2/rasterizer_storage_gles2.cpp [hunk 1]
```diff
@@ -5156,14 +5156,22 @@ void RasterizerStorageGLES2::render_target_set_external_texture(RID p_render_tar
...
-
-		// seeing we're rendering into this directly, better also use our depth buffer, just use our existing one :)
-		if (config.support_depth_texture) {
-			glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, rt->depth, 0);
+		// check if MSAA is active to set the correct depth buffer and target texture
+		if (rt->multisample_active) {
+			static const int msaa_value[] = { 0, 2, 4, 8, 16 };
+			int msaa = msaa_value[rt->msaa];
+			glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, rt->multisample_depth);
+			glFramebufferTexture2DMultisample(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, p_texture_id, 0, msaa);
```
Comment (akien-mga) [inline drivers/gles2/rasterizer_storage_gles2.cpp]: This only compiles on Android, other platforms don't fetch the extension for `glFramebufferTexture2DMultisample`.
Looking at other uses of this function, only Android seems to need using a texture for MSAA, other platforms use a renderbuffer directly.

Comment (NeoSpark314) [reply drivers/gles2/rasterizer_storage_gles2.cpp]: ok; so I can just check for `ANDROID_ENABLED` in this case?

**Comment by NeoSpark314** [comment]
I changed the PR to execute the new code only when `#ifdef ANDROID_ENABLED`

**Comment by clayjohn** [comment]
@BastiaanOlij Is there any chance that a desktop platform or iOS may use an ``external_texture``?

**Comment by BastiaanOlij** [comment]
@clayjohn the Oculus desktop driver uses the same approach and from what I know of OpenXR it also will require this logic.

The way these systems work is that they require a texture chain to be created into which the images are rendered. So instead of using the render target Godot creates we provide the target and it changes for every eye and every frame we render. The SDK takes the chain of images to do timewarp logic as I understand it and keep the framerate up by reprojecting the images. 

You can always do a copy from what Godot renders to these texture chains (that is how I did the original Oculus plugin) but that is very wasteful and a definite bottleneck on mobile hardware so we came up with this approach.

**Review by BastiaanOlij — COMMENTED** [review]
I don't think we can get way with only making this work on Android. This will need to work on desktop as well. Both Oculus Desktop and OpenXR use texture chains that require the plugin to setup the render target so they too will have issues under GLES2 when MSAA is enabled.

**Comment by NeoSpark314** [comment]
I now changed the PR to mirror the checks found in the function `void RasterizerStorageGLES2::_render_target_allocate(RenderTarget *rt)`. But I can't test if it works for Oculus Desktop and OpenXR .

File: drivers/gles2/rasterizer_storage_gles2.cpp [hunk 2]
```diff
@@ -5156,14 +5156,26 @@ void RasterizerStorageGLES2::render_target_set_external_texture(RID p_render_tar
...
-		// set our texture as the destination for our framebuffer
-		glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, p_texture_id, 0);
-
-		// seeing we're rendering into this directly, better also use our depth buffer, just use our existing one :)
-		if (config.support_depth_texture) {
-			glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, rt->depth, 0);
+		// check if MSAA is active to set the correct depth buffer and target texture for android
+		if (rt->multisample_active) {
+#if defined(GLES_OVER_GL) || defined(IPHONE_ENABLED)
+			glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, p_texture_id);
```
Comment (clayjohn) [inline drivers/gles2/rasterizer_storage_gles2.cpp]: I have a feeling that this isn't going to create an MSAA framebuffer. I would appreciate if someone could test though.

**Comment by akien-mga** [comment]
I agree with the comments in https://github.com/godotengine/godot/issues/33188#issuecomment-551227035 that we can already merge this to fix MSAA on mobile VR.

Desktop VR apparently needs to be handled partly in the plugins themselves, so we should open issue(s) to make sure to handle this before 3.2 in the various desktop VR plugins @BastiaanOlij @NeoSpark314.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**